presto-main/src/main/java/com/facebook/presto/metadata/SignatureBinder.java

```diff
@@ -0,0 +1,440 @@
+        if (!boundVariables.isPresent()) {
+            return Optional.empty();
+        }
+        return Optional.of(bindVariables(declaredSignature, boundVariables.get(), actualArgumentTypes.size()));
+    }
+
+    public Optional<BoundVariables> matchAndBindSignatureVariables(List<? extends Type> actualArgumentTypes)
+    {
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        BoundVariables.Builder boundVariablesBuilder = BoundVariables.builder();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+        BoundVariables boundVariables = boundVariablesBuilder.build();
+        checkAllTypeVariablesAreBound(boundVariables);
+        return Optional.of(boundVariables);
+    }
+
+    public Optional<BoundVariables> matchAndBindSignatureVariables(List<? extends Type> actualArgumentTypes, Type actualReturnType)
+    {
+        BoundVariables.Builder boundVariablesBuilder = BoundVariables.builder();
+
+        TypeSignature expectedReturnTypeSignature = declaredSignature.getReturnType();
+        if (!matchAndBindArgument(expectedReturnTypeSignature, actualReturnType, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+
+        List<TypeSignature> expectedArgumentSignatures = declaredSignature.getArgumentTypes();
+        boolean variableArity = declaredSignature.isVariableArity();
+        if (!matchArguments(expectedArgumentSignatures, actualArgumentTypes, variableArity, boundVariablesBuilder)) {
+            return Optional.empty();
+        }
+
+        calculateVariableValuesForLongConstraints(boundVariablesBuilder);
+
+        BoundVariables boundVariables = boundVariablesBuilder.build();
+        checkAllTypeVariablesAreBound(boundVariables);
+        return Optional.of(boundVariables);
+    }
+
+    private boolean matchArguments(
+            List<TypeSignature> expectedArgumentSignatures,
+            List<? extends Type> actualArgumentTypes,
+            boolean variableArity,
+            BoundVariables.Builder boundVariables)
+    {
+        if (variableArity) {
+            if (actualArgumentTypes.size() < expectedArgumentSignatures.size() - 1) {
+                return false;
+            }
+        }
+        else {
+            if (expectedArgumentSignatures.size() != actualArgumentTypes.size()) {
+                return false;
+            }
+        }
+
+        // Bind the variable arity argument first, to make sure it's bound to the common super type
+        if (variableArity && actualArgumentTypes.size() >= expectedArgumentSignatures.size()) {
+            List<? extends Type> variableArityParameterTypes = actualArgumentTypes.subList(
```

Comment:
The variable names are too long and overly verbose, and make the code hard to read. I'd rename these to:

```java
List<? extends Type> tail = actualTypes.subList(expectedTypes.size() - 1, actualTypes.size());
Optional<Type> commonType = typeManager.getCommonSuperType(tail);
if (!commonType.isPresent()) {
    return false;
}
TypeSignature expected = expectedTypes.get(expectedTypes.size() - 1);
if (!bind(expected, commonType.get(), variableBinder)) {
    return false;
}
```